### ISSUE TITLE
Null safety package for graphql

### ISSUE THREAD

**Issue by clarenceh** [zino-app/graphql-flutter#797]
Hi,

I am migrating my Flutter app to null safety packages and just want to know if there is any plan for this wonderful Graphql package in releasing a null safety version.

Currently I am using Firebase in my app and when I want to use a null safety version of firebase package, it depends on http_parser: ^4.0.0-nullsafety which graphql is not yet support.

I like GraphQL and Flutter and very happy to contribute if possible.

Cheers

**Comment by micimize** [comment]
This seems like it will be fairly involved, as we'll need to migrate all of the [gql project](https://github.com/gql-dart/gql/) we depend on as well. Hopefully the tooling is decent – this is definitely something we intend to / have to do.

Thankfully, dart's team is very considerate and null safety is backwards compatible, so the migration shouldn't be a total nightmare.

**Comment by micimize** [comment]
I have published the necessary gql packages: https://github.com/gql-dart/gql/pull/194.

Now we're waiting on `normalize`. and #817.

**Comment by 7mada123** [comment]
Hi, Any update for when null safety will be migrate to the package.

**Comment by b-jan** [comment]
We're waiting for the `normalize` package: https://github.com/gql-dart/ferry/issues/154

**Comment by 7mada123** [comment]
Thank you for keep supporting this amazing package ❤

**Comment by lypborges** [comment]
@micimize I think normalize now supports [null safety in this version](https://pub.dev/packages/normalize/versions/0.5.0-nullsafety.0)

**Comment by micimize** [comment]
@lypborges yes – in fact I have published a `4.0.3-alpha.1` that depends on null safe packages with help from @b-jan, so those looking to get ahead of the curve can depend on it.

**Comment by 7mada123** [comment]
I get an error when using a package that depend on uuid >= 3.0.0 because it's depend on crypto 3.0.0 and since the 4.0.3-alpha.1 version depend on uuid_enhanced ^3.0.2 which is depend on crypto ^2.0.0, I get the error because of crypto.
I don't know but is uuid_enhanced and uuid do the same thing? if so, maybe is it better to use uuid on the package since it has more support.

**Comment by lypborges** [comment]
@micimize tks for the response and great job on this project. I'm running ` dart pub outdated --mode=null-safety` and the message looks like it's not null safety yet. Any way I can help?

<img width="518" alt="Screen Shot 2021-03-06 at 16 00 18" src="https://user-images.githubusercontent.com/1749820/110217838-1dcaf800-7e95-11eb-824d-621a7a2ce3ef.png">

**Comment by micimize** [comment]
I just released [graphql 5.0.0-nullsafety.1](https://pub.dev/packages/graphql/versions/5.0.0-nullsafety.1) and  [graphql_flutter 5.0.0-nullsafety.1](https://pub.dev/packages/graphql_flutter/versions/5.0.0-nullsafety.1), and pushed the code to the [nullsafety](https://github.com/zino-app/graphql-flutter/tree/nullsafety) branch.

I squashed all the bugs our examples and tests surfaced, but I expect there to be a lot more. @lypborges what will help a lot is people using the prereleases and PRing fixes to `nullsafety` here as well as over at [gql](https://github.com/gql-dart/gql/tree/nullsafety).

These `nullsafety` packages should be considered the bleeding edge. I mostly blindly trusted `dart migrate` except for where it was breaking.

**Comment by psrcek** [comment]
Aren't we supposed to be able to return null in the AuthLink getToken method to indicate no authorization? Currently that isn't possible because of null safety.

I swear something like this used to work before `5.0.0-nullsafety.1`:

```dart
authLink = AuthLink(
  getToken: () async {
    final token = await getToken();
    return token == null ? null : 'Bearer $token';
  },
);
```

**Comment by micimize** [comment]
@psrcek I think you're right. I just made a [nullsafety](https://github.com/zino-app/graphql-flutter/labels/nullsafety) label – go ahead and open an issue and label it

**Comment by micimize** [comment]
@psrcek published `5.0.0-nullsafety.2` with a fix

**Comment by psrcek** [comment]
Are you sure you published? I can't seem to be able to use `5.0.0-nullsafety.2` in my pubspec.

**Comment by ol88** [comment]
@micimize @psrcek The fix in `5.0.0-nullsafety.2` did not work for me.
I might be missing something but I believe the type for `getToken` still doesn't allow null.

I made a pull request with a fix that's working for me, changing `getToken` type to allow nullable String to be returned: https://github.com/zino-app/graphql-flutter/pull/875

This might deserved to be moved to its own issue if more discussion needed.

**Comment by micimize** [comment]
@ol88 right you are. It was only allowing synchronous null. Merged and will try and release soon

**Comment by psrcek** [comment]
I don't know what your definition of soon is, but mine isn't 10 days... Could you release this so that we can use it?

**Comment by micimize** [comment]
@psrcek have had lots going on recently. You can always use [git dependencies](https://dart.dev/tools/pub/dependencies#git-packages) and `dependency_overrides` to depend on the branch directly

**Comment by psrcek** [comment]
That's exactly what I have been doing (what else could I do heh), but it's not really good to rely on git because it can change at any time without notice. I don't really understand your point though, what's so hard about publishing a package?

**Comment by micimize** [comment]
https://pub.dev/packages/graphql/versions/5.0.0-nullsafety.5 it is the same as https://pub.dev/packages/graphql/versions/5.0.0-beta.1, but should be preferred as the dependency as I believe prereleases are alphanumerically prioritized.
